File: leetcode17.cpp
class Solution {
public:
	string dic[10] = {{""},{""},{"abc"},{"def"},{"ghi"},{"jkl"},{"mno"},{"pqrs"},{"tuv"},{"wxyz"}};
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if(digits.size() == 0)
        	return res;
        int n = digits.size();
        string tmp(n, 0);
        dfs(tmp,0,n,res,digits);
        return res;
    }

    void dfs(string &tmp, int cur, int depth, vector<string> &res, string &digits)
    {
    	if(cur >= depth)
    	{
    		res.push_back(tmp);
    		return;
    	}
    	for (int i = 0; i < dic[digits[cur]-'0'].size(); ++i)
    	{
    		tmp[cur] = dic[digits[cur] - '0'][i];
    		dfs(tmp,cur+1,depth,res,digits);
    	}
    	return;
    }
};

File: leetcode17.java
public class Solution {
    private static final String[] refer = { "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz" };
    
    public List<String> letterCombinations(String digits) {
        List<String> res = new ArrayList<>();
        StringBuffer s = new StringBuffer();
        if(!digits.equals("")){
            backtrack(res,digits, s, 0);
            return res;
        }
        return res;
    }
    
    private void backtrack(List<String> res, String digits, StringBuffer s, int start){
        // end condition
        if(start == digits.length()){
            res.add(s.toString());
            return;
        }
        String tmp = refer[digits.charAt(start) - '0'];
        for(int i = 0; i < tmp.length(); i++){
            s.append(tmp.charAt(i));
            backtrack(res,digits,s,start+1);
            s.deleteCharAt(s.length() - 1);
        }
    }
}
